hnn_core/optimization/general_optimization.py

```diff
@@ -0,0 +1,409 @@
+
+
+class Optimizer:
+    def __init__(self, net, tstop, constraints, set_params, solver='bayesian',
+                 obj_fun='dipole_rmse', scale_factor=1.,
+                 smooth_window_len=None, max_iter=200):
+        """Parameter optimization.
+
+        Parameters
+        ----------
+        net : Network
+            The network object.
+        tstop : float
+            The simulated dipole's duration.
+        constraints : dict
+            The user-defined constraints.
+        set_params : func
+            User-defined function that sets parameters in network drives.
+        solver : string
+            The optimizer, 'bayesian' or 'cobyla'.
+        obj_fun : string
+            The objective function to be minimized.
+        scale_factor : float, optional
+            The dipole scale factor. The default is 1.
+        smooth_window_len : float, optional
+            The smooth window length. The default is None.
+        max_iter : int, optional
+            The max number of calls to the objective function. The default is
+            200.
+
+        Attributes
+        ----------
+        net : Network
+            The network object.
+        constraints : dict
+            The user-defined constraints.
+        max_iter : int
+            The max number of calls to the objective function.
+        solver : string
+            The optimizer, 'bayesian' or 'cobyla'.
+        obj_fun : func
+            The objective function to be minimized.
+        scale_factor : float
+            The dipole scale factor.
+        smooth_window_len : float
+            The smooth window length.
+        tstop : float
+            The simulated dipole's duration.
+        net_ : Network
+            The network object with optimized drives.
+        obj_ : list
+            The objective function values.
+        opt_params_ : list
+            The list of optimized parameter values.
+        """
+
+        if net.external_drives:
+            raise ValueError("The current Network instance has external " +
+                             "drives, provide a Network object with no " +
+                             "drives.")
+        self.net = net
+        self.constraints = constraints
+        self._set_params = set_params
+        self.max_iter = 200
+        # Optimizer method
+        if solver == 'bayesian':
+            self.solver = 'bayesian'
+            self._assemble_constraints = _assemble_constraints_bayesian
+            self._run_opt = _run_opt_bayesian
+        elif solver == 'cobyla':
+            self.solver = 'cobyla'
+            self._assemble_constraints = _assemble_constraints_cobyla
+            self._run_opt = _run_opt_cobyla
+        else:
+            raise ValueError("solver must be 'bayesian' or 'cobyla'")
+        # Response to be optimized
+        if obj_fun == 'dipole_rmse':
+            self.obj_fun = _rmse_evoked
+        else:
+            raise ValueError("obj_fun must be 'dipole_rmse'")
```

Comment:
So we're currently not allowing users to pass in their own callable here?

Reply:
Not at the moment, I think once we get this merged I can do that really easily

Reply:
Fair enough!